contractsV5/FeeHandler.sol

```diff
@@ -1,9 +1,205 @@
+        lastBurnBlock = block.number;
+        (, uint rewardInBPS, uint rebateInBPS, uint epoch, uint expiryBlock) = kyberDAOContract.getLatestBRRData();
+        brrAndEpochData = encodeData(rewardInBPS, rebateInBPS, epoch, expiryBlock);
+    }
+
+    modifier onlyDAO {
+        require(
+            msg.sender == address(kyberDAOContract),
+            "Only the DAO can call this function."
+        );
+        _;
+    }
+
+    modifier onlyKyberNetwork {
+        require(
+            msg.sender == address(kyberNetworkContract),
+            "Only the internal KyberNetwork contract can call this function."
+        );
+        _;
+    }
+
+    event AccumulateReserveRebate(
+        address[] eligibleWallets,
+        uint[] rebatePercentages,
+        uint totalRebateAmtWei,
+        uint totalRewardAmtWei,
+        uint totalBurnAmtWei
+    );
+
+    // Todo: future optimisation to accumulate rebates for 2 rebate wallet
+    // encode totals, 128 bits per reward / rebate
+    function handleFees(address[] calldata eligibleWallets, uint[] calldata rebatePercentages) external payable onlyKyberNetwork returns(bool) {
+        // Decoding BRR data
+        (uint rewardInBPS, uint rebateInBPS, uint epoch, uint expiryBlock) = decodeData();
+        uint burnInBPS = BPS - rewardInBPS - rebateInBPS;
+
+        // Check current block number
+        if(block.number > expiryBlock) {
+            (burnInBPS, rewardInBPS, rebateInBPS, epoch, expiryBlock) = kyberDAOContract.getLatestBRRData();
+
+            // Update brrAndEpochData
+            brrAndEpochData = encodeData(rewardInBPS, rebateInBPS, epoch, expiryBlock);
+        }
+
+        uint rebateWei = rebateInBPS * msg.value / BPS;
+        uint rewardWei = rewardInBPS * msg.value / BPS;
+        for(uint i = 0; i < eligibleWallets.length; i ++) {
+            // Internal accounting for rebates per reserve wallet (totalRebatesPerRebateWallet)
+            totalRebatesPerRebateWallet[eligibleWallets[i]] = rebateWei * rebatePercentages[i] / 100;
+        }
+
+        // Internal accounting for total rebates (totalRebates)
+        totalRebates += rebateWei;
+        // Internal accounting for reward per epoch (totalRewardsPerEpoch)
+        totalRewardsPerEpoch[epoch] += rewardWei;
+        // Internal accounting for total rewards (totalRewards)
+        totalRewards += rewardWei;
+
+        emit AccumulateReserveRebate(eligibleWallets, rebatePercentages, rebateWei, rewardWei, msg.value - rebateWei - rewardWei);
+
+        return true;
+    }
+
+    event DistributeRewards(address staker, uint amountWei);
+
+    function claimStakerReward(address staker, uint percentageInPrecision, uint epoch) public onlyDAO returns(uint) {
+        // Amount of reward to be sent to staker
+        uint amount = totalRewardsPerEpoch[epoch] * percentageInPrecision / PRECISION;
+
+        // Update total rewards and total rewards per epoch
+        totalRewardsPerEpoch[epoch] -= amount;
```

Comment:
need to make sure this isn't under flowing

Reply:
done